File: nprgrab.pl
#!/usr/bin/perl
use Net::SMTP;
use File::Copy;
use MIME::Lite;

$usage = "Usage: $0 <fa|we|me|atc|ww> [#]\n where # is number of days ago\n";

if(!@ARGV){
	print $usage;
	exit(2);
}
chomp($prg = $ARGV[0]);
if($prg !~ /fa|we|me|atc|ww/){
	print $usage;
	print "($prg)\n";
	exit(2);
}
$prog = $prg;
#if($prg eq "me"){
#	$prog = "morning";
#}
$npath = "/home/elyons/radio";
$tpath = "/var/tmp";
$plist = "/var/www/html/" . $prg . ".m3u";
%tits = ('fa'=>'Fresh Aire', 'wesat'=>'Weakened Edition Saturday', 'wesun'=>'Weakened Edition Sunday', 'me'=>'Morning Edition', 'atc'=>'All Things Considered', 'ww'=>'Wait Wait Dont Tell Me');
chomp($year = `/bin/date +\%Y`);
chomp($d = `/bin/date +\%m\%d`);
chomp($w = `/bin/date +\%w`);
$dstr = "$year$d";
$mark = "/var/tmp/" . $prg . $d;
if($ARGV[1]){
	chomp($off = $ARGV[1]);
	if($off !~ /^[0-9]+$/){
		print $usage;
		print "($off)\n";
		exit(2);
	}
	$dcmd = "/bin/date -d \"$off days ago\" +\%m\%d";
	$wcmd = "/bin/date -d \"$off days ago\" +\%w";
	$ycmd = "/bin/date -d \"$off days ago\" +\%Y";
	chomp($d = `$dcmd`);
	chomp($w = `$wcmd`);
	chomp($year = `$ycmd`);
	$mark = "/var/tmp/" . $prg . $d;
}
else{
	if((-e "$mark")&&(!$ENV{TERM})){
# need to work on this...
#		@dif = `diff /var/tmp/$prg.m3u /var/www/html/$prg.m3u`;
#		if($? == 0){
			&logg("job already completed for $prg$d");
			exit();
#		}
#		foreach $l (@diff){
#			next unless($l =~ /^</);
#			$l =~ s/^< //;
#			&logg("playlist differs from previous: $l");
#			open(ADDS, ">>$plist");
#			print ADDS $l;
#			close(ADDS);
#		}
	}
	else{
		&logg("flag file $mark not found, proceeding");
	}
	if(-e "$plist"){
		if(-z "$plist"){
			&logg("playlist file $plist has no size, deleting");
			unlink($plist);
			exit();
		}
		$mtime = (stat("$plist"))[9];
#		$age = strftime("%H", localtime($mtime));
#		$hour = strftime("%H", localtime(time));
		$hour = time;
		$diff = $hour - $mtime;
		if($diff > 28800){
			&logg("playlist file $plist is stale, aborting");
			unlink($plist);
			exit(3);
		}
	}
	else{
		&logg("playlist file $plist is absent");
		exit();
	}
}
if($prg eq "we"){
	if($w == 6){
		$prg = "wesat";
	}
	elsif($w == 0){
		$prg = "wesun";
	}
	else{
		print "Error, not a weekend: d = ".$d.", w = ".$w." \n";
		exit(2);
	}
}
chdir($tpath);
$get = "/usr/bin/wget -q -T 30 -O " . $prg . "\.m3u http://lumpy/" . $prg;
if($ARGV[1]){
#	chomp($plist = $ARGV[2]);
#	$get = "/usr/bin/wget -q -T 30 -O " . $prg . "\.m3u http://lumpy/" . $prg . "-" . $d . "\.m3u"; 
	$get .= "-" . $d; 
	if($ENV{TERM}){
		print "$get\n";
	}
}
$get .= "\.m3u";
if(system($get)!=0){
	&logg("wget failed for playlist $prog");
	if($ENV{TERM}){print "URL: $get\n";}
	exit(1);
}
my $got = $prog . "\.m3u";
if(! -e "$got"){
	&logg("playlist $got not found after wget");
	exit(1);
}
$done=0;
foreach $url (`cat $got`){
#	my $song = "Unknown Title";
	chomp($url);
	if($url =~ /^\# (.+)$/){
		$song = $1;
		$song =~ s/\'//g;
		$song =~ s/ Or / or /g;
		$song =~ s/ Of / of /g;
		$song =~ s/ And / and /g;
		$song =~ s/ The / the /g;
		next;
	}
	# http://public.npr.org/anon.npr-mp3/npr/me/2011/01/20110106_me_14.mp3
	elsif($url =~ /^http.+\.mp3/){
		$fname = $1 . "_" . $prg . "_" . $2 . "\.mp3";
		$i++;
		$t = $i;
		if($i < 10){
			$t = "0" . $i;
		}
		$oname = $prg . $d . "t" . $t . "\.mp3";
		if(-e "$npath/$oname"){
			&logg("File $oname already exists, skipping.");
			next;
		}
		else{
			if($ENV{TERM}){
				print "Grabbing " . $prg . $d . " track $t...\n";
			}
#			if(system("/usr/bin/wget -q $url")!=0){
			if(system("/usr/bin/wget -qO $fname $url")!=0){
				&logg("wget failed for $url: $!");
				next;
			}
			unless(-e "$tpath/$fname"){
				if(system("/bin/mv $tpath/$fname* $tpath/$fname")!=0){
					&logg("move failed for $fname: $!");
					next;
				}
			}
			$recode = "/usr/bin/lame -a -m m -b 24 --resample 16 --scale 2 --quiet --add-id3v2 --ta NPR";
			$recode .= " --tl \"".$d." ".$tits{$prg}. "\"";
			if($song =~ /.+/){
				$recode .= " --tt \"".$song."\"";
			}
			else{
				$recode .= " --tt \"".$tits{$prg}." ".$t."\"";
			}
			$recode .= " --ty \"$year\"";
			$recode .= " --tn \"$t\"";
			$recode .= " --tg \"Speech\"";
			$oname =~ s/^wes[atun]+/we/;
			if(-e "$npath/$oname"){
				print "File $oname already retrieved, skipping\n";
				next;
			}
			if(system("$recode $tpath/$fname $npath/$oname 2>&1 >/dev/null")==0){
	#			&logg("Re-encoded and tagged file $fname");
				$done++;
				push(@flies, $oname);
				unlink("$tpath/$fname") or &logg("Debug: unlink returned false on $tpath/$fname: $!");
			}
			else{
				&logg("Warning, lame re-encoding failed on $fname");
			#	&lamer($fname);
			}
		}
#		push(@flies, $fname);
	}
	else{
		&logg("no regex match on $url");
	}
}
if($done < 1){
	&logg("no files available after wgets");
	exit(1);
}
else{
	if(! -e "$mark"){
		system("echo $done > $mark");
	}
	$done .= " file";
	if($done != 1){
		$done .= "s";
	}
	&logg("Re-encoded and tagged $done");
	$repor = "Re-encoded and tagged $done for $prog\n";
	chdir($npath);
}
$prg =~ s/^wes[atun]+/we/;
@brony = `rsync -auv $npath/$prg$d* 192.168.0.43:Music/mp3/ 2>\&1`;
if($?!=0){
	&logg("rsync to brony at wired address failed");
	$repor .= "rsync to brony at wired address failed\n";
	@brony = `rsync -auv $npath/$prg$d* 192.168.0.9:Music/mp3/ 2>\&1`;
	if($?!=0){
		&logg("rsync to brony at wifi address failed, aborting");
		$repor .= "rsync to brony at wifi address failed, aborting\n";
	}
}
if(grep(/mp3/, @brony)){
	@sent = grep(/\.mp3$/, @brony);
	$dunn = scalar(@sent);
	&logg("Copied $dunn files to brony for $prg");
	$repor .= "Copied $dunn files to brony for $prg\n";
	open(SENT, ">/var/tmp/went2brony");
	foreach $b (@brony){
		next if($b !~ /mp3/);
		chomp($b);
		$b =~ s/^([^\s]+)\s+.*$/$1/;
		print SENT "$b\n";
	}
	close(SENT);
}
else{
	&logg("No files were copied to brony for $prg");
	$repor .= "No files were copied to brony for $prg\n";
}
&domail($repor);
print "seem to have reached the end.\n" if($ENV{TERM});
exit();
sub logg( "$" ){
	chomp($tim = `date +\"\%b \%d \%T\"`);
#	$tim = localtime(time);
	$out = "@_";
	$lfile = "/home/elyons/log/nprgrab";
	open(LOG, ">>$lfile") or die "Can't open logfile $lfile: $!\n";
	select LOG unless($ENV{TERM});
	print "[$tim] (nprgrab) $out\n";
	close(LOG);
}
sub domail( "$" ){
	$tim = localtime(time);
	chomp($host = `hostname`);
	$subject = "Files available for $prog";
	$bod= "@_";
	&logg($bod);
	$bod .= "\nMessage generated $tim by $host:$0\n";
	$msg = MIME::Lite->new(
	#	SetSender=>'rserling@pantload.net',
	#	Return-Path =>'rserling@pantload.net',
	#	From    =>'rserling@pantload.net',
		SetSender=>'wolfstools@gmail.com',
		Return-Path =>'wolfstools@gmail.com',
		From    =>'wolfstools@gmail.com',
		To      =>'bitshag@gmail.com',
		Subject =>"$subject",
		Type    =>'text/plain',
		Data =>"$bod"
	);
	$msg->send;
}

File: nprgrab.py
#!/usr/bin/env python3
import os
import time
import sys
import subprocess
from datetime import datetime, timedelta
import smtplib
from email.mime.text import MIMEText
import shutil

# Constants
TITLES = {
    'atc': 'All Things Considered',
    'fa': 'Fresh Aire',
    'me': 'Morning Edition',
    'we': 'Weakened Edition',
    'wesat': 'Weakened Edition Saturday', 
    'wesun': 'Weakened Edition Sunday',
    'ww': 'Wait Wait Dont Tell Me'
}

BASE = os.getenv("HOME")
LPATH = os.path.join(BASE, "Music")
NPATH = os.path.join(LPATH, "radio")
TPATH = "/var/tmp"

def log_message(message):
    """Log messages with timestamp"""
    timestamp = datetime.now().strftime("%b %d %H:%M:%S")
    log_file = os.path.join(LPATH, "nprgrab.log")

    log_entry = f"[{timestamp}] (nprgrab) {message}\n"
    if 'TERM' in os.environ:
        print(log_entry, end='')

    with open(log_file, 'a') as f:
        f.write(log_entry)

def send_email(subject, body):
    """Send email notification"""
    hostname = subprocess.getoutput('hostname')
    timestamp = datetime.now().strftime("%c")
    body += f"\nMessage generated {timestamp} by {hostname}:{sys.argv[0]}\n"
    
    msg = MIMEText(body)
    msg['Subject'] = subject
    msg['From'] = os.getenv('SMTP_FROM', 'wolfstools@gmail.com')
    msg['To'] = 'bitshag@gmail.com'
    
    try:
        with smtplib.SMTP('localhost') as s:
            s.send_message(msg)
        return True
    except Exception as e:
        log_message(f"ERROR: Email send failed: {str(e)}")
        return False

def main():
    usage = f"Usage: {sys.argv[0]} <atc|fa|me|we|wesat|wesun|ww> [#]\n where # is number of days ago\n"
    
    if len(sys.argv) < 2:
        print(usage)
        sys.exit(2)
        
    prg = sys.argv[1]
    if prg not in ['fa', 'we', 'wesat', 'wesun', 'me', 'atc', 'ww']:
        print(usage)
        print(f"({prg})")
        sys.exit(2)
        
    prog = prg
    
    # Get current date or date offset
    if len(sys.argv) > 2:
        try:
            days_ago = int(sys.argv[2])
            target_date = datetime.now() - timedelta(days=days_ago)
        except ValueError:
            print(usage)
            print(f"({sys.argv[2]})")
            sys.exit(2)
    else:
        target_date = datetime.now()
        
    year = target_date.strftime("%Y")
    date_str = target_date.strftime("%m%d")
    weekday = target_date.strftime("%w")
    
    # Handle weakened edition special case
    if prg == "we":
        if weekday == "6":
            prg = "wesat"
        elif weekday == "0":
            prg = "wesun"
        else:
            print(f"Error, not a weekend: d = {date_str}, w = {weekday}")
            sys.exit(2)
            
    mark_file = os.path.join(TPATH, f"{prg}{date_str}")
    playlist = f"/var/www/html/{prog}.m3u"
    
    # Check if already processed
    if os.path.exists(mark_file):
        log_message(f"WARN: job already completed for {prg}{date_str}")
        sys.exit(0)
        
    # Process playlist
    if not os.path.exists(playlist):
        log_message(f"ERROR: playlist file {playlist} is absent")
        sys.exit(0)
        
    if os.path.getsize(playlist) == 0:
        log_message(f"ERROR: playlist file {playlist} has no size, deleting")
        os.unlink(playlist)
        sys.exit(0)
        
    # Check playlist staleness, frogot use case
    mtime = os.path.getmtime(playlist)
    if (time.time() - mtime) > 28800:  # 8 hours
        log_message(f"WARN: playlist file {playlist} is stale, aborting")
        os.unlink(playlist)
        sys.exit(3)
        
    # Process MP3 files
    os.chdir(TPATH)
    done = 0
    files_processed = []
    
    with open(playlist) as f:
        story = "Unknown Title"
        track_num = 0
        
        for line in f:
            line = line.strip()
            
            # Handle title lines
            if line.startswith('# '):
                new_story = line[2:]
                if new_story == story:
                    continue
                    
                story = new_story
                story = story.replace("'", "")
                for word in ['Or', 'Of', 'And', 'The', 'For', 'A', 'An']:
                    story = story.replace(f" {word} ", f" {word.lower()} ")
                continue
                
            # Handle MP3 URLs
            if line.endswith('.mp3') and line.startswith('http'):
                track_num += 1
                track_str = f"{track_num:02d}"
                output_name = f"{prg}{date_str}t{track_str}.mp3"
                
                if os.path.exists(os.path.join(NPATH, output_name)):
                    log_message(f"WARN: File {output_name} already exists, skipping.")
                    done += 1
                    continue
                    
                if 'TERM' in os.environ:
                    print(f"Grabbing {prg}{date_str} track {track_str}...")
                    
                # Download and process file
                try:
                    subprocess.run(['wget', '-qO', output_name, line], check=True)
                    
                    # Build LAME command
                    lame_cmd = [
                        'lame', '-a', '-m', 'm', '-b', '24', '--resample', '16',
                        '--scale', '2', '--quiet', '--add-id3v2',
                        '--ta', 'NPR',
                        '--tl', f"{date_str} {TITLES[prg]}",
                        '--ty', year,
                        '--tn', track_str,
                        '--tg', 'Speech'
                    ]
                    
                    if story != "Unknown Title":
                        lame_cmd.extend(['--tt', story])
                        
                    lame_cmd.extend([output_name, os.path.join(NPATH, output_name)])
                    
                    subprocess.run(lame_cmd, check=True, capture_output=True)
                    
                    done += 1
                    files_processed.append(output_name)
                    os.unlink(output_name)
                    
                except subprocess.CalledProcessError as e:
                    log_message(f"ERROR: Error processing {output_name}: {str(e)}")
                    continue
                    
    if done < 1:
        log_message("ERROR: no files available after downloads")
        sys.exit(1)
    else:
        if not os.path.exists(mark_file):
            with open(mark_file, 'w') as f:
                f.write(str(done))
                
        files_text = "file" if done == 1 else "files"
        log_message(f"INFO: Re-encoded and tagged {done} {files_text} for {prog}")
        
        os.chdir(NPATH)
        log_message("INFO: seem to have reached the end")

if __name__ == "__main__":
    main()
